fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/view/ViewExpression.java

```diff
@@ -0,0 +1,197 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * An abstraction that represents a stream of values that can be represented as tuple-encoding keys and binary values
+ * in FoundationDB.
+ *
+ * <p>
+ * Visually, a view expression is quite similar to a query in a declarative language like SQL. A view expression is
+ * composed of several clauses, including:
+ * </p>
+ * <ul>
+ *     <li>A root {@link Source}, which defines the stream of values from which the expression's {@link Element}s are generated.</li>
+ *     <li>A projection, which is a collection of {@link Element}s that generate values.</li>
+ *     <li>
+ *         A sort order, which is an ordered list of {@link Element}s that define the order of the stream. By definition,
+ *         the sort order is the lexicographic order of the elements when packed together as tuple.
+ *       </li>
+ *     <li>A predicate, in the form of a {@link QueryPredicate}</li>
+ * </ul>
+ *
+ * <p>
+ * Because view expressions are quite general but are closely tied to the semantics of the tuple layer, they are very
+ * useful for planning tasks such as incremental index selection. To see how an index is represented as a view
+ * expression, see {@link #fromIndexDefinition(String, Collection, KeyExpression)}.
+ * </p>
+ */
+public class ViewExpression {
+    @Nonnull
+    private final List<Element> select;
+    @Nullable
+    private final QueryPredicate predicate;
+    @Nonnull
+    private final Source rootSource;
+    @Nonnull
+    private final List<Element> orderBy;
+
+    public ViewExpression(@Nonnull List<Element> select, @Nullable QueryPredicate predicate, @Nonnull Source rootSource, @Nonnull List<Element> orderBy) {
+        this.select = select;
+        this.predicate = predicate;
+        this.rootSource = rootSource;
+        this.orderBy = orderBy;
+    }
+
+    @Nonnull
+    public List<Element> getOrderBy() {
+        return orderBy;
+    }
+
+    /**
+     * Generate a view expression that represents the structure of an index defined using a {@link KeyExpression}.
+     * @param indexType the type of the index as specified in the meta-data
+     * @param recordTypes the set of record types used to generate the index
+     * @param rootExpression the root key expression of the index as specified in the meta-data
+     * @return a view expression representing the given index structure
+     */
+    public static ViewExpression fromIndexDefinition(@Nonnull String indexType,
+                                                     @Nonnull Collection<String> recordTypes,
+                                                     @Nonnull KeyExpression rootExpression) {
+        ViewExpression.Builder builder = ViewExpression.builder();
+        if (indexType.equals(IndexTypes.VALUE)) {
```

Comment:
Agreed/

I don't think we want to pollute the Issue space with all of these (I have more than a hundred notes on this sort of thing, the majority of which will probably never be resolved because a design decision will obviate them).